src/profile/hooks/useProfileDataEditor.ts

```diff
@@ -0,0 +1,34 @@
+import { useContext, useMemo } from 'react';
+
+import { ProfileContext } from '../context/ProfileContext';
+import {
+  EditData,
+  EditDataType,
+  createEditorForDataType,
+} from '../helpers/editData';
+
+export type ProfileDataEditorReturnType = {
+  editDataList: EditData[];
+};
+
+export function useProfileDataEditor({
+  dataType,
+}: {
+  dataType: EditDataType;
+}): ProfileDataEditorReturnType {
+  const { data: profileData } = useContext(ProfileContext);
+  if (!profileData) {
+    throw new Error(
+      'No profile data for useProfileDataEditor. Hook used before data has been fetched.'
+    );
+  }
+  const { getEditData } = useMemo(
+    () => createEditorForDataType(profileData, dataType),
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    []
+  );
```

Comment:
Not in this case. "createEditorForDataType" should never be run again, because it initialises editor data. Currently there is no way to update profile, but in the future updates are done separately.

Reply:
So, this hook (`useProfileDataEditor`) implementation makes assumptions, that the caller never changes the `dataType` argument, and neither does the profile data change? I can't say if that's a common way to implement hooks, but feels weird to me.

Reply:
Yes, dataType never changes in the same component that uses the hook. Same instance of the component never render first "emails" and then "phones". 

And initialiser (createEditorForDataType) should not be called twice. Unsaved user input would be destroyed. In the layout user may have added a new phone number and 3 others numbers edited, but not saved yet. So saving one and re-creating data, would destroy unsaved ones.

 Each Component (and dataType) uses their own instance of the hook. So it's like creating a class instance on component initialisation. It is encapsulated to that component.

Reply:
Is it this hook's job to say whether the caller changes the `dataType` argument or not? This hook shouldn't care in which kinds of components it's used. The components using this hook are probably responsible for handling data updates properly and if that includes keeping `dataType` constant, then a component is free to do that.

If this **hook** really can't handle changing `dataType` argument, then shouldn't it e.g. throw an exception in case `dataType` changes instead of just not acting on it in any way?

Btw. I've been reading [this long discussion](https://github.com/facebook/create-react-app/issues/6880) about missing dependencies and seems that most of the cases of having empty dependency lists have been debunked.

Reply:
> Is it this hook's job to say whether the caller changes the dataType argument or not? 

Yes, because the hook has only one purpose. It is not for universal use.

>This hook shouldn't care in which kinds of components it's used. 

It doesn't care. It works on any component without errors. It just prevents re-initialisation the data the hook is using or user's data would be destroyed.

Of course, I can add

```
useMemo(()=>{
 throw new Error("DataType changed")
},[dataType]).
```

It will never trigger, but developer can see that it is not supported - if cant be seen from empty dependency array.

It is empty because it should never be run again and it's much easier to realise it when seeing an empty array. 

There is no situation in the UI where component would switch dataTypes. The component is made for one fixed dataType which is obvious from usage. 

If one day there is a new dataType switcher, the component would not use a hook which purpose is to **keep** data of single dataType.

Reply:
The empty dependency list is an implementation detail. The future developer might only check the interface of the hook. They see this:

```typescript
function useProfileDataEditor({
  dataType,
}: {
  dataType: EditDataType;
}): ProfileDataEditorReturnType
```

A function which takes in an `EditDataType` and returns a `ProfileDataEditorReturnType`. Nothing in this interface indicates that the user should always use the same value for the `dataType` argument.

This applies to any programming interface: the API should be intuitive (and sometimes the implementation isn't even available to be read). In this regard I'm thinking that hooks are just functions. Am I thinking this somehow incorrectly?

I'm probably over thinking this now, but could there be a "hook factory"? Something like this:

```typescript
function getUseProfileDataEditor({
  dataType,
}: {
  dataType: EditDataType;
}): () => ProfileDataEditorReturnType {
  return (): ProfileDataEditorReturnType => {
    // The hook implementation would be here...
  };
}
```

That would be used something like this then:

```typescript
// Create hook function somewhere
const useProfileDataEditor = getUseProfileDataEditor({dataType: 'basic-data'});
// ...
// Then use it somewhere...
const { editDataList } = useProfileDataEditor();
```

That would prevent passing different `dataType` arguments to the hook. I'm not sure what should be in the dependency list within the hook in this case :thinking:

Reply:
Hooks are not just functions that return a value. They were created so functional components can have state that do not change every time the function is called - which is on every render.

From the docs: "With Hooks, you can extract stateful logic from a component " 

Like React's own **useState**(initialValue), the **useProfileDataEditor**(dataType) returns the same value every time even if the arguments change. Same logic there  - once called they do not change on next calls. They keep the state. 

Hook's inner state is controlled with what ever they return in the function call. useState returns its value and a function to change it. If useProfileDataEditor would allow a new dataType, it would return a function for that. React developers who use hooks must know these rules.

>Nothing in this interface indicates that the user should always use the same value for the dataType argument.
 
Datatype does not have to be the same every time. It just has no effect after the first call. Like in useState(). Hooks are hooks, and not just functions, for a reason - they don't mutate/re-init when called. 

Hooks interface really cannot tell you that what a hook does internally. Interface of useState: 
`function useState<S>(initialState: S | (() => S)): [S, Dispatch<SetStateAction<S>>];`
Return value is important part of the hook's interface.

One main concept with hooks is that they cannot be called conditionally. These will result in compiler error

if(dataType === "phones"){
  usePhonesDataEditor(....)
}

if(!initialData){
 [initialData] = useState(createEditorForDataType(dataType))
}

Second concept is that they must be called inside a component. For these reasons factory functions would not work. 
There could be usePhoneDataEditor, useEmailDataEditor etc. But then also PhoneDataEditor etc that uses the hook.
Singletons or Redux could be other solutions. Or create new contexts: /PhoneEditorDataContext, EmailEditorDataContext....

Reply:
Thanks for a thorough explanation -- and reminder, I have read these before, but it seems to require some repetition until I fully get it. The analogy with `useState` makes sense.